src/vma/sock/sockinfo_tcp.cpp

```diff
@@ -2011,6 +2011,18 @@ int sockinfo_tcp::bind(const sockaddr *__addr, socklen_t __addrlen)
 		return -1; //error
 	}
 
+	if (!m_bound.is_anyaddr() && !m_bound.is_local_loopback()) {
+		ip_address local_ip(m_bound.get_in_addr());
+
+		net_device_resources_t* p_nd_resources = get_nd_resources((const ip_address)local_ip);
+		if (p_nd_resources) {
+			/* Ring is available now but not ready to receive */
+			m_p_rx_ring = p_nd_resources->p_ring;
```

Comment:
Issue 1
=====================================================

Here you only execute:
```
m_p_rx_ring = p_nd_resources->p_ring;
```
However in sockinfo::attach_receiver() you do more with p_nd_resources, such as:
1. Incrementing reference count
2. Inserting into m_rx_flow_map[]
See code sample:
```
	// Now we have the net_device object (created or found)
	p_nd_resources->refcnt++;   // <==================== Ophir: please look here

	// Map flow in local map
	m_rx_flow_map[flow_key] = p_nd_resources->p_ring; // <======================= Ophir: please look here

	// Save the new CQ from ring
	rx_add_ring_cb(flow_key, p_nd_resources->p_ring);

	// Attach tuple
	BULLSEYE_EXCLUDE_BLOCK_START
	unlock_rx_q();
	if (!p_nd_resources->p_ring->attach_flow(flow_key, this)) {
		lock_rx_q();
		si_logerr("Failed to attach %s to ring %p", flow_key.to_str(), p_nd_resources->p_ring);
		return false;
	}
	lock_rx_q();
	BULLSEYE_EXCLUDE_BLOCK_END

	// Registered as receiver successfully
	si_logdbg("Attached %s to ring %p", flow_key.to_str(), p_nd_resources->p_ring);
```
Please justify why you are NOT doing those operations under sockinfo_tcp::bind()
Better keeping consistent logic

Issue 2
=====================================================

I am thinking of a situation where you get one ring during bind() then your ring changes during connect().
Confirmation to this behavior is pending on tests you are currently running.
If this is the situation - we may get errors during ring changes and we should check for it and behave correctly.

Looking forward to your tests results regarding possible ring changes between bind() and connect().